src/common/KokkosKernels_Sorting.hpp

```diff
@@ -0,0 +1,631 @@
+    //sort 4 bits at a time, into 16 buckets
+    ValueType mask = 0xF;
+    //maskPos counts the low bit index of mask (0, 4, 8, ...)
+    Ordinal maskPos = 0;
+    for(Ordinal s = 0; s < (sortBits + 3) / 4; s++)
+    {
+      //Count the number of elements in each bucket
+      Ordinal count[16] = {0};
+      Ordinal offset[17];
+      if(!inAux)
+      {
+        for(Ordinal i = 0; i < n; i++)
+        {
+          count[(values[i] & mask) >> maskPos]++;
+        }
+      }
+      else
+      {
+        for(Ordinal i = 0; i < n; i++)
+        {
+          count[(valuesAux[i] & mask) >> maskPos]++;
+        }
+      }
+      offset[0] = 0;
+      //get offset as the prefix sum for count
+      for(Ordinal i = 0; i < 16; i++)
+      {
+        offset[i + 1] = offset[i] + count[i];
+      }
+      //now for each element in [lo, hi), move it to its offset in the other buffer
+      //this branch should be ok because whichBuf is the same on all threads
+      if(!inAux)
+      {
+        //copy from *Over to *Aux
+        for(Ordinal i = 0; i < n; i++)
+        {
+          Ordinal bucket = (values[i] & mask) >> maskPos;
+          valuesAux[offset[bucket + 1] - count[bucket]] = values[i];
+          permAux[offset[bucket + 1] - count[bucket]] = perm[i];
+          count[bucket]--;
+        }
+      }
+      else
+      {
+        //copy from *Aux to *Over
+        for(Ordinal i = 0; i < n; i++)
+        {
+          Ordinal bucket = (valuesAux[i] & mask) >> maskPos;
+          values[offset[bucket + 1] - count[bucket]] = valuesAux[i];
+          perm[offset[bucket + 1] - count[bucket]] = permAux[i];
+          count[bucket]--;
+        }
+      }
+      inAux = !inAux;
+      mask = mask << 4;
+      maskPos += 4;
+    }
+  });
+  //Move values back into main array if they are currently in aux.
+  //This is the case if an odd number of rounds were done.
+  if(((sortBits + 3) / 4) % 2 == 1)
+  {
+    Kokkos::parallel_for(Kokkos::ThreadVectorRange(thread, n),
+    [=](Ordinal i)
+    {
+      values[i] = valuesAux[i];
+      perm[i] = permAux[i];
+    });
+  }
+  if(minVal != 0)
+  {
+    //remove bias to restore original values
+    Kokkos::parallel_for(Kokkos::ThreadVectorRange(thread, n),
+    [=](Ordinal i)
+    {
+      values[i] += minVal;
+    });
+  }
+}
+
+template<typename Value>
+struct DefaultComparator
+{
+  KOKKOS_INLINE_FUNCTION bool operator()(const Value lhs, const Value rhs) const
+  {
+    return lhs < rhs;
+  }
+};
+
+//Bitonic merge sort (requires only comparison operators and trivially-copyable)
+//Pros: In-place, plenty of parallelism for GPUs, and memory references are coalesced
+//Con: O(n log^2(n)) serial time is bad on CPUs
+//Good diagram of the algorithm at https://en.wikipedia.org/wiki/Bitonic_sorter
+template<typename Ordinal, typename ValueType, typename TeamMember, typename Comparator = DefaultComparator<ValueType>>
+KOKKOS_INLINE_FUNCTION void
+bitonicSortTeam(ValueType* values, Ordinal n, const TeamMember mem)
```

Comment:
You went with team as a suffix. Let us make sure we are consistent

Reply:
I will add "Thread" as a suffix to radix then, that seems the most clear.